src/GafferScene/SetExpression.cpp

```diff
@@ -0,0 +1,276 @@
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferScene/SetExpression.h"
+#include "GafferScene/PathMatcherData.h"
+
+#include <boost/spirit/include/qi.hpp>
+
+using namespace Gaffer;
+using namespace GafferScene;
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+
+#include <boost/config/warning_disable.hpp>
+#include <boost/spirit/include/qi.hpp>
+#include <boost/variant/recursive_variant.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/spirit/include/phoenix_operator.hpp>
+#include <boost/spirit/include/phoenix_function.hpp>
+
+#include <iostream>
+#include <vector>
+#include <string>
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+
+// AST
+struct binary_op;
+struct nil {};
+
+struct ExpressionAst
+{
+	typedef
+	boost::variant<
+		nil, // can't happen!
+		std::string,
+		boost::recursive_wrapper<ExpressionAst>,
+		boost::recursive_wrapper<binary_op>
+		>
+	type;
+
+	ExpressionAst()
+		: expr(nil()) {}
+
+	template <typename Expr>
+	ExpressionAst(Expr const& expr)
+		: expr(expr) {}
+
+	ExpressionAst& operator+=(ExpressionAst const& rhs);
+	ExpressionAst& operator-=(ExpressionAst const& rhs);
+
+	type expr;
+};
+struct binary_op
+{
+	binary_op(
+		char op,
+		ExpressionAst const& left,
+		ExpressionAst const& right)
+		: op(op), left(left), right(right) {}
+
+	char op;
+	ExpressionAst left;
+	ExpressionAst right;
+};
+
+ExpressionAst& ExpressionAst::operator+=(ExpressionAst const& rhs)
+{
+	expr = binary_op('|', expr, rhs);
+	return *this;
+}
+
+ExpressionAst& ExpressionAst::operator-=(ExpressionAst const& rhs)
+{
+	expr = binary_op('&', expr, rhs);
+	return *this;
+}
+
+// Visiting the AST
+// ----------------
+struct AstPrinter
+{
+typedef void result_type;
+
+void operator()(qi::info::nil) const {}
+void operator()(std::string n) const { std::cout << n; }
+
+void operator()(ExpressionAst const& ast) const
+{
+	boost::apply_visitor(*this, ast.expr);
+}
+
+void operator()(binary_op const& expr) const
+{
+	std::cout << "op:" << expr.op << "(";
+	boost::apply_visitor(*this, expr.left.expr);
+	std::cout << ", ";
+	boost::apply_visitor(*this, expr.right.expr);
+	std::cout << ')';
+}
+
+};
+
+struct AstEvaluator
+{
+typedef PathMatcher result_type;
+
+AstEvaluator( const ScenePlug *scene )
+{
+	m_scene = scene;
+}
+
+result_type operator()(std::string setName) const
+{
+	// TODO: handle case in which the setName given is not a valid set
```

Comment:
Mh ok. I guess I had a case in mind where a user would add an expression like this:

`(mySetA & mSetB)`  _Note the typo_

What you're proposing is to just silently return an empty PathMatcher, right? Would you be opposed to at least printing a warning that the requested set couldn't be found? What I had in mind would look like this:

```
result_type operator()( std::string setName ) const
{
  const std::vector<IECore::InternedString> &setNames = m_scene->setNames()->readable();
  if( std::find( setNames.begin(), setNames.end(), IECore::InternedString( setName ) ) != setNames.end() )
  {
    return m_scene->set( setName )->readable();
  }
  else
  {
    throw IECore::Exception( boost::str( boost::format( "Set %s is not available in SetExpression." ) % setName ) );
  }
}
```